contracts/leaser/src/migrate.rs

```diff
@@ -2,101 +2,305 @@ use lease::api::MigrateMsg;
+#[cfg_attr(test, derive(Debug, Eq, PartialEq))]
+pub struct MigrationResult {
+    pub msgs: Batch,
+    pub next_customer: Option<Addr>,
+}
+
+/// Builds a batch of messages for the migration of up to `max_leases`
+///
+/// Consumes the customers iterator to the next customer or error.
+pub fn migrate_leases<I, LI>(
+    mut customers: I,
+    lease_code_id: u64,
+    max_leases: NbInstances,
+) -> ContractResult<MigrationResult>
 where
-    I: Iterator<Item = ContractResult<Addr>>,
+    I: Iterator<Item = ContractResult<Customer<LI>>>,
+    LI: ExactSizeIterator<Item = Addr>,
 {
-    let no_msgs = MigrateBatch::new(lease_code_id);
-    let migrated_msgs = leases.fold(no_msgs, MigrateBatch::migrate_lease);
-    migrated_msgs.try_into()
+    let mut msgs = MigrateBatch::new(lease_code_id, max_leases);
+    customers
+        .find_map(|may_customer| match may_customer {
+            Ok(customer) => msgs.migrate_or_be_next(customer),
+            Err(err) => Some(Err(err)),
+        })
+        .transpose()
+        .map(|next_customer| MigrationResult {
+            msgs: msgs.into(),
+            next_customer,
+        })
+}
+
+impl<LeaseIter> Customer<LeaseIter>
+where
+    LeaseIter: Iterator<Item = Addr>,
+{
+    pub fn from(customer: Addr, leases: LeaseIter) -> Self {
+        Self { customer, leases }
+    }
+}
+
+impl MigrationResult {
+    pub fn try_add_msgs<F>(mut self, add_fn: F) -> ContractResult<Self>
+    where
+        F: FnOnce(&mut Batch) -> ContractResult<()>,
+    {
+        add_fn(&mut self.msgs).map(|()| self)
+    }
 }
 
 struct MigrateBatch {
     new_code_id: u64,
-    may_batch: ContractResult<Batch>,
+    leases_left: NbInstances,
+    msgs: Batch,
 }
 impl MigrateBatch {
-    fn new(new_code_id: u64) -> Self {
+    fn new(new_code_id: u64, max_leases: NbInstances) -> Self {
         Self {
             new_code_id,
-            may_batch: Ok(Batch::default()),
+            leases_left: max_leases,
+            msgs: Default::default(),
         }
     }
 
-    fn migrate_lease(mut self, lease_contract: ContractResult<Addr>) -> Self {
-        let op = |mut batch: Batch| {
-            lease_contract.and_then(|lease| {
-                batch
-                    .schedule_migrate_wasm_no_reply(&lease, MigrateMsg {}, self.new_code_id)
-                    .map(|_| batch)
-                    .map_err(Into::into)
-            })
-        };
-
-        self.may_batch = self.may_batch.and_then(op);
-        self
+    /// None if there is enough room for all customer's leases, otherwise return the customer
+    fn migrate_or_be_next<LI>(&mut self, mut customer: Customer<LI>) -> Option<ContractResult<Addr>>
+    where
+        LI: ExactSizeIterator<Item = Addr>,
+    {
+        let may_leases_nb: Result<NbInstances, _> = customer.leases.len().try_into();
+        match may_leases_nb {
+            Err(err) => Some(Err(err.into())),
+            Ok(leases_nb) => {
+                if leases_nb <= self.leases_left {
```

Comment:
What happens if that is never the case and the gas would otherwise not be enough/hits a limit? In my opinion previous commit's way was better in that it did deal with this case too.

Reply:
the previous commit in this PR was buggy - it mixed customer with lease addresses